File: src/main/java/ar/com/flow/bankaccount/transaction/withdrawal/SufficientFunds.java
package ar.com.flow.bankaccount.transaction.withdrawal;

import ar.com.flow.bankaccount.BankAccount;
import ar.com.flow.bankaccount.transaction.Preconditions;
import ar.com.flow.money.InsufficientFundsException;
import ar.com.flow.money.Money;

public class SufficientFunds implements Preconditions {
    private final BankAccount account;
    private final Money amount;
    private final WithdrawalLimit withdrawalLimit;

    public SufficientFunds(BankAccount account, Money amount) {
        this.account = account;
        this.amount = amount;
        this.withdrawalLimit = account.getWithdrawalLimit();
    }

    public void check() {
        if (!withdrawalLimit.accepts(account, amount)) {
            throw new InsufficientFundsException();
        }
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/domain/withdrawal/SufficientFunds.kt
package ar.com.flow.bankaccount.domain.withdrawal

import ar.com.flow.bankaccount.domain.BankAccount
import ar.com.flow.bankaccount.domain.transaction.Precondition
import ar.com.flow.money.InsufficientFundsException
import ar.com.flow.money.Money

class SufficientFunds(private val account: BankAccount, private val amount: Money) : Precondition {
    override fun check() {
        if (!account.allowsWithdraw(amount)) {
            throw InsufficientFundsException()
        }
    }
}